### PR TITLE
VSR: `op_repair_min()` + chronological repair

### PR DESCRIPTION
**Pull request by sentientwaffle**
## `op_repair_min()` + chronological repair

(copied from the first commit's message)

This commit has two different changes because I implemented them at the same time and they were too much work to untangle. Sorry!

1. Repair prepares in chronological order.
2. Repair prepares back to `op_repair_min()` which is:
  - For backups, it is usually `op_checkpoint + 1`.
  - For primary, it is usually `op_head - journal_slots + pipeline`.

Guiding principle: State transfer is expensive, prefer to repair WAL whenever possible.

### 1. Repair prepares in chronological order.

(Formerly prepares were repaired in _reverse_ chronological order, i.e. starting at op-head, and then going backwards.)

#### Primary:

For a *primary* (during a view-change), the order of repairs is irrelevant because:

  - no one else is advancing their ops anyway,
  - and we need to repair these prepares to progress.

#### Backup:

(Background:
- The WAL is an on-disk ring buffer.
- A "slot" is `message.header.op % journal_slot_count`.
- `journal_slot_count` is a cluster constant.)

But for a *backup* (which just joined a view), the oldest prepares will be overwritten by the primary (and other backups) earlier than the newest prepares.
Consider a backup's WAL (`journal_slot_count = 8`):

           slot  0  1  2  3  4  5  6  7
    primary WAL  8  9 10 11 12 13  6  7
     backup WAL  0  1  2  3  4  5  _  _

In this example, the backup is lagging behind the primary — it is missing ops 6…13. It must repair to catch up.
But while it is repairing, the primary (and the rest of the cluster) are preparing even more ops (14, 15, ...).
The backup needs to repair op 6 as soon as possible since it is about to be overwritten (by op 14, because 14%8=6).

If the backup *can't* repair the prepare (already overwritten) it will have to resort to state transfer.

Tl;dr: To maximize the chances of repairing a prepare, repair prepares from oldest to newest.

(Aside: As a bonus, chronological repair enables better pipelining of repair + commit — we can commit op=X while repairing op=X+1).

### 2. Repair prepares back to `op_repair_min()`.

(See `op_repair_min()`'s comment & code for details.)

At a high level, the idea is:

  - The backup needs to get caught up to the primary as soon as possible.
    - Repair any messages we need to progress.
    - And repair any messages we would need to recover if we crash before checkpointing.
  - The primary should repair more aggressively (i.e. further back):
    - The further back the primary can repair, the more it can help repair backups.
    - This decreases the likelihood that backups will need to resort to state transfer to catch up.

#### Primary:

For the primary, the reason this _isn't_ as simple as `op_repair_min = op_head -| (journal_slot_count - 1)` is that `op_head` can venture ahead and then backtrack (by as much as `pipeline_prepare_queue_max`) when a view-change truncates uncommitted ops.
We also know that the "venture ahead" is bounded by the checkpoint trigger.
When a primary is starting a view & repairing, all committed ops greater-or-equal-to the primary's `op_repair_min()` are guaranteed to be available in _someone's_ WAL (assuming ≤f/(2f+1) faults).

## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

### CODE REVIEW

File: src/vsr.zig [hunk 1]
```diff
@@ -236,13 +236,16 @@ pub const Header = extern struct {
...
+    ///
+    /// * A `request_headers` sets this to the maximum op requested (inclusive).
     op: u64 = 0,
 
     /// The commit number of the latest committed prepare. Committed ops are immutable.
     ///
     /// * A `do_view_change` sets this to `commit_min`, to indicate the sending replica's progress.
     ///   The sending replica may continue to commit after sending the DVC.
     /// * A `start_view` sets this to `commit_max`.
+    /// * A `request_headers` sets this to the minimum op requested (inclusive).
```
Comment (sentientwaffle) [inline src/vsr.zig:248]: These haven't changed, just adding documentation.

File: src/vsr.zig [hunk 2]
```diff
@@ -1336,19 +1339,18 @@ const ViewChangeHeadersArray = struct {
...
+                        assert(!current.log_view_primary);
                         break;
                     },
                     // Breaks are due to:
-                    // - entries before checkpoint, which are not repaired
+                    // - entries before checkpoint, which are not repaired by backups
                     .chain_break => {
-                        assert(op <= options.op_checkpoint);
+                        assert(!current.log_view_primary and op <= options.op_checkpoint);
                         break;
```
Comment (sentientwaffle) [inline src/vsr.zig:1354]: I have a follow-up PR for this one (wip) that will remove this whole function — `op_repair_min()` doesn't actually make view-change header generation any easier, but it did show me a simpler implementation.

**Comment by sentientwaffle** [comment]
Note that although this PR allows primaries to repair farther back in the prepare history, lagging replicas cannot yet take advantage of this extra history to catch up. That will be in a follow-up PR which is not yet ready.

File: src/vsr/replica.zig [hunk 3]
```diff
@@ -3926,43 +3977,36 @@ pub fn ReplicaType(
             }
 
             // Request any missing or disconnected headers:
-            if (self.commit_min != self.op) {
```
Comment (matklad) [inline src/vsr/replica.zig:3929]: Could you adjust the comment on `fn repair` to describe the overall flow. I find myself a bit lost in which order we repair things in. I _think_ what happens is that: 

* we request the latest prepare `if (self.op < commit_max_limit)` to learn the hash of the latest commited op
* we request missing headers in reverse-chronological order
* we reuqest missing prepares in chronological order

Comment (sentientwaffle) [reply src/vsr/replica.zig:3929]: Adding:

```
        /// Repair. Each step happens in sequence — step n+1 executes when step n is done.
        ///
        /// 1. Advance the head op to `op_repair_max = min(op_checkpoint_trigger, commit_max)`.
        ///    To advance the head op we request+await a SV. Either:
        ///    - the SV's "hook" headers include op_checkpoint_trigger (if we are ≤1 wrap behind), or
        ///    - the SV is too far ahead, so we will fall back from WAL repair to state transfer.
        /// 2. Acquire missing or disconnected headers in reverse chronological order, backwards from
        ///    op_repair_max.
        ///    A header is disconnected if it breaks the chain with its newer neighbor to the right.
        /// 3. Repair missing or corrupt prepares in chronological order.
        ///    If we are the primary, this may result in nack/truncation before we start the view.
        /// 4. Commit up to op_repair_max. If committing triggers a checkpoint, op_repair_max
        ///    increases, so go to step 1 and repeat.

```

Comment (sentientwaffle) [reply src/vsr/replica.zig:3929]: Ah, just realized that the reference to the "hook" headers refers to code that isn't in this PR :facepalm: 
Ignore that for now.

File: src/vsr/replica.zig [hunk 4]
```diff
@@ -3926,43 +3986,36 @@ pub fn ReplicaType(
...
-                            .view = self.view,
-                            .commit = range.op_min,
-                            .op = range.op_max,
-                        });
-                    }
-                    return;
+                assert(range.op_min >= self.op_repair_min());
+                assert(range.op_max < self.op);
+
+                if (self.choose_any_other_replica()) |replica| {
```
Comment (matklad) [inline src/vsr/replica.zig]: What happens if there's no other replica? It feels a bit weird that wee do nothing for that case --- feels like we should maybe shutdown here if sole_replica() ?

Comment (sentientwaffle) [reply src/vsr/replica.zig]: :+1: I think this can be `self.choose_any_other_replica().?` since solo replicas cannot tolerate corruption anyhow.

File: src/vsr/replica.zig [hunk 5]
```diff
@@ -4390,58 +4429,34 @@ pub fn ReplicaType(
...
-                        continue;
-                    }
-
-                    // If this is an uncommitted op, and we are the primary in `view_change` status,
+                    // This is an uncommitted op — if we are the primary in `view_change` status,
                     // then we will `request_prepare` from the cluster, set `nack_prepare_op`,
                     // and stop repairing any further prepares:
                     // This will also rebroadcast any `request_prepare` every `repair_timeout` tick.
                     if (self.repair_prepare(op)) {
                         if (self.nack_prepare_op) |nack_prepare_op| {
```
Comment (matklad) [inline src/vsr/replica.zig:4455]: This I think is another benefit of chronological repair! We `nack` the earliest prepare, so we only have one nacking round (with reverse-chronologicial, I think we might end up nacking a bunch of preps one-by-one?)

File: src/vsr/replica.zig (hunk 4 above)
Comment (matklad) [reply src/vsr/replica.zig]: Hm, I think the same would be true for another call to this function, so perhaps it can just assert(!sole_replica()) rather than return a null?

File: src/vsr/replica.zig [hunk 6]
```diff
@@ -3719,6 +3719,57 @@ pub fn ReplicaType(
...
+        ///
+        /// - Repairing uncommitted ops is necessary as primary — we will need them to proceed.
+        /// - Repairing committed + not-yet-checkpointed ops is useful because we might crash,
+        ///   in which case we will need them to recover.
+        /// - Repairing committed + checkpointed ops:
+        ///   - backups don't repair checkpointed ops, since there is no guarantee that they will
+        ///     ever be available (if our head is behind), and we don't want to stall the new view
+        ///     startup.
+        ///   - primaries do repair checkpointed ops — as many as are guaranteed to exist anywhere in
+        ///     the cluster, so that they can help lagging backups catch up.
```
Comment (matklad) [inline src/vsr/replica.zig:3736]: Hm, I don't understand this bit: why primary can repair beyond checkpoint, but a backup cant. :thinking: or maybe I do?

If we are a backup, we don't know the cluster's state. The cluster might have moved ahead of us, so preparess before checkpoint might not be available.

If we are the primary, we know the cluster's state! The only way cluster can move ahead is if there's a view change (but then we'll view jump and no longer be primary).

So, basically, when the primary repairs, it is guaranteed to either repair or view jump.

If a backup repairs, in general it _might_ get stuck if prepare it requests is no longer available. This might happen even if we request very few prepares, but that's OK, as we'll switch to state transfer if repair gets stuck.

In other words, primary's repair always suceeds, so primary tries to request **as much as we could** (so, journal-full)

Backup repair might fail, so we try to request as **little as we must**

File: src/vsr/replica.zig [hunk 7]
```diff
@@ -3719,6 +3719,57 @@ pub fn ReplicaType(
...
+                        // And we don't need op_checkpoint's prepare/header in our log if we crash.
+                        break :op self.op_checkpoint() + 1;
+                    }
+                }
+            };
+
+            assert(op <= self.op);
+            assert(op <= self.commit_min + 1);
+            assert(op <= self.op_checkpoint() + 1);
+            assert(op >= self.op -| (constants.journal_slot_count - 1));
```
Comment (matklad) [inline src/vsr/replica.zig]: micro nit, but maybe phrase this in terms of `self.op - op` -- the lenght of the rage of ops we are repairing?

File: src/vsr/replica.zig (hunk 6 above)
Comment (matklad) [reply src/vsr/replica.zig:3736]: > since there is no guarantee that they will ever be available

Hm, this is not the whole story, right? This is not only about availability, but also about correcness, as per

```
// It is critical that we do not repair checkpointed ops; their slots now belong
```

?

Comment (sentientwaffle) [reply src/vsr/replica.zig:3736]: > If we are a backup, we don't know the cluster's state. The cluster might have moved ahead of us, so preparess before checkpoint might not be available.
> If we are the primary, we know the cluster's state! The only way cluster can move ahead is if there's a view change (but then we'll view jump and no longer be primary).

This is exactly right! The backups know the _past_ state of the primary, but if the primary has advanced since then, they might be requesting ops that are unavailable. The primary knows its state right now.

> Hm, this is not the whole story, right? This is not only about availability, but also about correcness, as per

That's a good point, I'll update the comment. If backups repaired before the checkpoint they could overwrite slots which belong to the current wrap, which could lead to the backup incorrectly nacking a message.

File: src/vsr/replica.zig [hunk 8]
```diff
@@ -3868,10 +3922,19 @@ pub fn ReplicaType(
             };
         }
 
-        /// Starting from the latest journal entry, backfill any missing or disconnected headers.
-        /// A header is disconnected if it breaks the chain with its newer neighbor to the right.
-        /// Since we work back from the latest entry, we should always be able to fix the chain.
-        /// Once headers are connected, backfill any dirty or faulty prepares.
+        /// Repair. Each step happens in sequence — step n+1 executes when step n is done.
+        ///
+        /// 1. Advance the head op to `op_repair_max = min(op_checkpoint_trigger, commit_max)`.
```
Comment (matklad) [inline src/vsr/replica.zig:3927]: Maybe rename `commit_max_limit` to `op_repair_max` in the body of fn?

Comment (matklad) [reply src/vsr/replica.zig:3927]: Or is it from the follow up PR?

Comment (sentientwaffle) [reply src/vsr/replica.zig:3927]: It is already called `op_repair_max` in the follow-up PR — it ends up being refactored into a (very short) function since it gets reused elsewhere too.

File: src/vsr/replica.zig [hunk 9]
```diff
@@ -3868,10 +3922,19 @@ pub fn ReplicaType(
...
+        ///    To advance the head op we request+await a SV. Either:
+        ///    - the SV's "hook" headers include op_checkpoint_trigger (if we are ≤1 wrap behind), or
+        ///    - the SV is too far ahead, so we will fall back from WAL repair to state transfer.
+        /// 2. Acquire missing or disconnected headers in reverse chronological order, backwards from
+        ///    op_repair_max.
```
Comment (matklad) [inline src/vsr/replica.zig:3932]: Hm, we go backwards from `self.op`. It feels like `op` might actually be higher than `op_repair_max`?

Comment (sentientwaffle) [reply src/vsr/replica.zig:3932]: It isn't clear here, but in the next PR we will enforce:
`op` never exceeds `op_repair_max`.

That is, `op <= op_checkpoint_trigger` — even if we learn of a higher op, we can't increase our head op beyond the trigger.

Comment (sentientwaffle) [reply src/vsr/replica.zig:3932]: `commit_max` however can go as high as it likes

File: src/vsr/replica.zig [hunk 10]
```diff
@@ -4138,12 +4188,9 @@ pub fn ReplicaType(
...
 
-            if (header.op <= self.commit_min) {
-                if (self.journal.header_with_op(header.op)) |existing| {
-                    // If we already committed this op, the repair must be the identical message.
-                    assert(header.checksum == existing.checksum);
-                }
-            }
+            // If we already committed this op, the repair must be the identical message.
+            assert(header.op <= self.op_checkpoint() or header.op > self.commit_min or
+                self.journal.has(header));
```
Comment (matklad) [inline src/vsr/replica.zig]: A bit hard to untangle assert. I'd maybe preserve an if here? 

```zig
if (header.op <= self.commit_min and header.op > self.op_checkpoint()) {
    assert(self.journal.has(header));
}
```

reeealy miss the implication operator here!

Comment (sentientwaffle) [reply src/vsr/replica.zig]: :+1: Good point, that is a bit much. Fwiw, it is actually

```
            if (self.op_checkpoint() < header.op and header.op <= self.commit_min) {
                assert(self.journal.has(header));
            }
```

The intent is: if it is committed but _not_ checkpointed, then we must already know the correct header.
(If it is checkpointed, then we might not have the correct header b/c we state transferred up to that checkpoint & don't have anything before it).

Comment (matklad) [reply src/vsr/replica.zig]: > Fwiw, it is actually

I _think_ I wrote the same thing, but skipped my own `lo < foo and foo < hi` rule, making it less readable.

File: src/vsr/replica.zig [hunk 11]
```diff
@@ -4621,29 +4662,21 @@ pub fn ReplicaType(
...
-                        // WAL wrap.
-                        return;
-                    } else {
-                        // If an op is committed but not checkpointed, we must still have the header.
-                        @panic("missing committed, uncheckpointed header");
-                    }
-                }
-            }
+            // If we already committed this op, the repair must be the identical message.
+            assert(header.op <= self.op_checkpoint() or header.op > self.commit_min or
```
Comment (matklad) [inline src/vsr/replica.zig]: same assert -> if here

**Review by matklad — APPROVED** [review]
